### PR TITLE
fix: Do not run DataFusion optimizer pass twice

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
Closes https://github.com/influxdata/influxdb_iox/issues/4800

One of the fluxtests is failing with
```
	multi_measure_negation...fail: query error: rpc error: code = InvalidArgument desc = Error running series plans for database '54425547585a6420_dfc574f9716be050': Schema error: Schema contains duplicate unqualified field name 'IsNull-Column-swap.host'
```

The underlying issues appears to be that one of the DataFusion optimizer passes messes up the plan when run twice,(tracked by https://github.com/apache/arrow-datafusion/issues/2712)

I believe the reason IOx is optimizing the plan twice is that the APIs to run / optimize `LogicalPlans` have been changed significantly since this code was first written, and now `SessionContext::create_physical_plan` also optimizes the plan.

# Changes
1. Work around the issue by avoiding the double optimization
2. Rename `IOxSessonContext::prepare_plan` to `IOxSessonContext::create_physical_plan` to mirror DataFusion `SessionContext`

# Testing
- [x] Verified that when run with this fix, the `multi_measure_negation` test in fluxtest no longer errors

```
cd /Users/alamb/Software/idpe && FLUXTEST_IOX_GRPC_ADDR=localhost:8082 FLUXTEST_IOX_HTTP_ADDR=localhost:8080 FLUXTEST_STORAGE_MODE=iox FLUXTEST_IOX_USE_EXISTING=1 go run ./internal/cmd/fluxtest -v --test multi_measure_negation
{"level":"warn","ts":"2022-06-08T14:07:32.856-0400","caller":"auth/store.go:1220","msg":"simple token is not cryptographically signed"}
{"level":"warn","ts":"2022-06-08T14:07:32.857-0400","caller":"etcdserver/metrics.go:224","msg":"failed to get file descriptor usage","error":"cannot get FDUsage on darwin"}
multi_measure_negation...success
```

### CODE REVIEW

File: compactor/src/compact.rs [hunk 1]
```diff
@@ -744,7 +744,7 @@ impl Compactor {
 
         let ctx = self.exec.new_context(ExecutorType::Reorg);
         let physical_plan = ctx
-            .prepare_plan(&plan)
+            .create_physical_plan(&plan)
```
Comment (alamb) [inline compactor/src/compact.rs:747]: I renamed this function to mirror what the underlying DataFusion API has evolved to

File: iox_query/src/exec/context.rs [hunk 2]
```diff
@@ -279,22 +279,16 @@ impl IOxSessionContext {
...
 
     /// Prepare (optimize + plan) a pre-created logical plan for execution
-    pub async fn prepare_plan(&self, plan: &LogicalPlan) -> Result<Arc<dyn ExecutionPlan>> {
-        let mut ctx = self.child_ctx("prepare_plan");
+    pub async fn create_physical_plan(&self, plan: &LogicalPlan) -> Result<Arc<dyn ExecutionPlan>> {
+        let mut ctx = self.child_ctx("create_physical_plan");
         debug!(text=%plan.display_indent_schema(), "prepare_plan: initial plan");
+        let physical_plan = ctx.inner.create_physical_plan(plan).await?;
 
-        let plan = ctx.inner.optimize(plan)?;
```
Comment (alamb) [inline iox_query/src/exec/context.rs:290]: The fix is to not call `optimize` here. It is called as part of `ctx.inner.create_physical_plan(plan)`: 

https://github.com/apache/arrow-datafusion/blob/1152c14f5a6bb6fa949d5162380f47aeeeca6696/datafusion/core/src/execution/context.rs#L886

Which calls 

https://github.com/apache/arrow-datafusion/blob/1152c14f5a6bb6fa949d5162380f47aeeeca6696/datafusion/core/src/execution/context.rs#L1341-L1350

File: query_tests/src/influxrpc/read_filter.rs [hunk 3]
```diff
@@ -750,6 +745,29 @@ async fn test_read_filter_on_field_single_measurement() {
...
+        "Series tags={_measurement=attributes, _field=color}\n  StringPoints timestamps: [8000], values: [\"blue\"]",
+        "Series tags={_measurement=cpu_load_short, host=server01, _field=value}\n  FloatPoints timestamps: [1000], values: [27.99]",
+        "Series tags={_measurement=cpu_load_short, host=server01, region=us-east, _field=value}\n  FloatPoints timestamps: [3000], values: [1234567.891011]",
+        "Series tags={_measurement=cpu_load_short, host=server01, region=us-west, _field=value}\n  FloatPoints timestamps: [0, 4000], values: [0.64, 3e-6]",
+        "Series tags={_measurement=status, _field=active}\n  BooleanPoints timestamps: [7000], values: [true]",
+        "Series tags={_measurement=swap, host=server01, name=disk0, _field=in}\n  FloatPoints timestamps: [6000], values: [3.0]",
+        "Series tags={_measurement=swap, host=server01, name=disk0, _field=out}\n  FloatPoints timestamps: [6000], values: [4.0]",
+    ];
+
+    run_read_filter_test_case(EndToEndTest {}, predicate, expected_results).await;
```
Comment (alamb) [inline query_tests/src/influxrpc/read_filter.rs:768]: this test errors (with the same error as in #4800) if `optimize()` is run twice

File: iox_query/src/exec/context.rs [hunk 4]
```diff
@@ -279,22 +279,16 @@ impl IOxSessionContext {
...
+        ctx.create_physical_plan(&logical_plan).await
     }
 
     /// Prepare (optimize + plan) a pre-created logical plan for execution
-    pub async fn prepare_plan(&self, plan: &LogicalPlan) -> Result<Arc<dyn ExecutionPlan>> {
-        let mut ctx = self.child_ctx("prepare_plan");
+    pub async fn create_physical_plan(&self, plan: &LogicalPlan) -> Result<Arc<dyn ExecutionPlan>> {
+        let mut ctx = self.child_ctx("create_physical_plan");
         debug!(text=%plan.display_indent_schema(), "prepare_plan: initial plan");
+        let physical_plan = ctx.inner.create_physical_plan(plan).await?;
```
Comment (NGA-TRAN) [inline iox_query/src/exec/context.rs:289]: Should we also replace `prepare_plan` in the debug info the line above, too?

File: query_tests/src/influxrpc/read_filter.rs [hunk 5]
```diff
@@ -187,12 +187,7 @@ async fn test_read_filter_invalid_predicate_case() {
         // https://github.com/influxdata/influxdb_iox/issues/3635
         // model what happens when a field is treated like a tag
         // CASE WHEN system" IS NULL THEN '' ELSE system END = 5;
-        .add_expr(
-            when(col("system").is_null(), lit(""))
-                .otherwise(col("system"))
-                .unwrap()
-                .eq(lit(5i32)),
-        );
+        .add_expr(make_empty_tag_ref_expr("system").eq(lit(5i32)));
```
Comment (NGA-TRAN) [inline query_tests/src/influxrpc/read_filter.rs:190]: 👍

**Review by NGA-TRAN — APPROVED** [review]
Makes sense

File: iox_query/src/exec/context.rs (hunk 4 above)
Comment (alamb) [reply iox_query/src/exec/context.rs:289]: good idea -- will do

Comment (alamb) [reply iox_query/src/exec/context.rs:289]: in https://github.com/influxdata/influxdb_iox/pull/4809/commits/22aef85911d42eb5c581f1661bc54014e61a9142

**Pull request merged**